File: api/Validators/journal.js
'use strict';

/** api/Validators/journal.js — Journal voucher (create + list). */

const Joi = require('joi');

const STATUSES = ['pending_tally', 'sent_to_tally', 'created', 'failed'];

const VCH_TYPES = ['Journal', 'Contra', 'Credit Note', 'Debit Note'];

const createJournalSchema = Joi.object({
    vch_type:     Joi.string().valid(...VCH_TYPES).default('Journal'),
    journal_date: Joi.string().trim().max(20).required(),
    dr_ledger:    Joi.string().trim().min(1).max(191).required(),
    cr_ledger:    Joi.string().trim().min(1).max(191).required(),
    amount:       Joi.number().positive().required(),
    narration:    Joi.string().trim().max(500).allow('', null),
});

const listJournalSchema = Joi.object({
    search:   Joi.string().trim().max(191).allow('', null),
    status:   Joi.string().valid(...STATUSES),
    page:     Joi.number().integer().min(1).default(1),
    per_page: Joi.number().integer().min(1).max(100).default(20),
});

module.exports = { createJournalSchema, listJournalSchema, STATUSES, VCH_TYPES };

File: app/lib/data/models/journal.dart
/// One journal voucher (Dr/Cr two-ledger entry) from `GET /api/v1/journals`.
/// pg returns numeric/bigint columns as strings, so coercions are defensive.
class Journal {
  const Journal({
    required this.id,
    required this.voucherNo,
    this.vchType,
    this.journalDate,
    this.drLedger,
    this.crLedger,
    this.amount,
    this.narration,
    this.status,
    this.createdAt,
  });

  final int id;
  final String voucherNo;
  final String? vchType; // Journal | Contra | Credit Note | Debit Note
  final String? journalDate;
  final String? drLedger;
  final String? crLedger;
  final num? amount;
  final String? narration;
  final String? status;
  final String? createdAt;

  factory Journal.fromJson(Map<String, dynamic> j) => Journal(
        id: _toInt(j['id']) ?? 0,
        voucherNo: _s(j['voucher_no']),
        vchType: _sn(j['vch_type']),
        journalDate: _sn(j['journal_date']),
        drLedger: _sn(j['dr_ledger']),
        crLedger: _sn(j['cr_ledger']),
        amount: _toNum(j['amount']),
        narration: _sn(j['narration']),
        status: _sn(j['status']),
        createdAt: _sn(j['created_at']),
      );

  static String _s(Object? v) => v == null ? '' : v.toString();
  static String? _sn(Object? v) {
    if (v == null) return null;
    final s = v.toString().trim();
    return s.isEmpty ? null : s;
  }

  static int? _toInt(Object? v) {
    if (v == null) return null;
    if (v is num) return v.toInt();
    final s = v.toString().trim();
    return s.isEmpty ? null : int.tryParse(s);
  }

  static num? _toNum(Object? v) {
    if (v == null) return null;
    if (v is num) return v;
    final s = v.toString().trim();
    return s.isEmpty ? null : num.tryParse(s);
  }
}
